src/ucp/tag/rndv.c

```diff
@@ -455,22 +455,92 @@ static ucs_status_t ucp_rndv_progress_send(uct_pending_req_t *self)
     return status;
 }
 
+static void ucp_rndv_zcopy_req_complete(ucp_request_t *req)
+{
+    ucp_request_send_buffer_dereg(req, ucp_ep_get_am_lane(req->send.ep));
+    ucp_request_complete_send(req, UCS_OK);
+}
+
+static void ucp_rndv_contig_zcopy_completion(uct_completion_t *self,
+                                             ucs_status_t status)
+{
+    ucp_request_t *sreq = ucs_container_of(self, ucp_request_t, send.uct_comp);
+    ucp_rndv_zcopy_req_complete(sreq);
+}
+
+static ucs_status_t ucp_rndv_zcopy_single(uct_pending_req_t *self)
+{
+    ucp_request_t *sreq = ucs_container_of(self, ucp_request_t, send.uct);
+    ucp_rndv_data_hdr_t hdr;
+
+    hdr.rreq_ptr = sreq->send.proto.rreq_ptr;
+
+    return ucp_do_am_zcopy_single(self, UCP_AM_ID_RNDV_DATA_LAST, &hdr, sizeof(hdr),
+                                  ucp_rndv_zcopy_req_complete);
+}
+
+static ucs_status_t ucp_rndv_zcopy_multi(uct_pending_req_t *self)
+{
+    ucp_request_t *sreq = ucs_container_of(self, ucp_request_t, send.uct);
+    ucp_rndv_data_hdr_t hdr;
+
+    hdr.rreq_ptr = sreq->send.proto.rreq_ptr;
+
+    return ucp_do_am_zcopy_multi(self,
+                                 UCP_AM_ID_RNDV_DATA,
+                                 UCP_AM_ID_RNDV_DATA,
+                                 UCP_AM_ID_RNDV_DATA_LAST,
+                                 &hdr, sizeof(hdr),
+                                 &hdr, sizeof(hdr),
+                                 ucp_rndv_zcopy_req_complete);
+}
+
 static ucs_status_t
 ucp_rndv_rtr_handler(void *arg, void *data, size_t length, void *desc)
 {
     ucp_rndv_rtr_hdr_t *rndv_rtr_hdr = data;
     ucp_request_t *sreq = (ucp_request_t*) rndv_rtr_hdr->sreq_ptr;
+    ucp_ep_h ep = sreq->send.ep;
 
     /* make sure that the ep on which the rtr was received on is connected */
-    ucs_assert_always(!ucp_ep_is_stub(sreq->send.ep));
+    ucs_assert_always(!ucp_ep_is_stub(ep));
     ucs_trace_req("RTR received. start sending on sreq %p", sreq);
 
-    /* dereg the original send request since we are going to send with bcopy */
-    ucp_request_send_buffer_dereg(sreq, ucp_ep_get_rndv_get_lane(sreq->send.ep));
+    if (sreq->send.length >= ucp_ep_config(ep)->zcopy_thresh) {
+        /* send with zcopy */
+        size_t max_zcopy;
```

Comment:
@shssf This is ANSI C compliant, since the declaration is the fist line inside {}.